src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -5669,7 +5669,25 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
           case LookupInaccessible(symbol, msg) => issue(AccessError(tree, symbol, context, msg))
           case LookupNotFound =>
             asTypeName orElse inEmptyPackage orElse lookupInRoot(name) match {
-              case NoSymbol => issue(SymbolNotFoundError(tree, name, context.owner, startContext, mode.in(all = PATTERNmode, none = APPSELmode | TYPEPATmode)))
+              case NoSymbol =>
+                def hasOutput = settings.outputDirs.getSingleOutput.isDefined || settings.outputDirs.outputs.nonEmpty
+                val hidden = !unit.isJava && name.isTypeName && hasOutput && {
+                  startContext.lookupSymbol(name.toTermName, _ => true) match {
+                    case LookupSucceeded(qualifier, symbol) if symbol.owner.hasPackageFlag && symbol.sourceFile != null =>
+                      symbol.owner.info.decls.lookupAll(name).toList match {
```

Comment:
This I don't immediately understand; lookup for `name` failed (we're about to issue an error), but `lookupSymbol(name.toTermName).owner.info.decls.lookupAll(name)` gives a result. How is that?

Reply:
The package is loaded from class path first, which creates symbols of existing class + its module & module class. Then compilation of source takes over the existing module symbol, which is entered into scope.

It's fishy that the package class has a set of members but not all are in scope. Previously I've read a comment about stale symbols from previous run; symbol history, like human history, is a story of wreckage and detritus. Somehow I had never fully considered the state of the environment during incremental compilation.